### ISSUE TITLE
Upgrade to Reactor 2.0.6 to fix performance regression in web sockets

### ISSUE THREAD

**Issue by krm1312** [spring-projects/spring-boot#4003]
Tried upgrading our 1.2.X spring boot application to 1.3.0.M5 today.  Was able to get all functionality to work, however, the web socket (publishing) performance has degraded to the point where it is not usable.  I kind of doubt this is boot related, but, since I was trying to upgrade boot and get all the new versions together, thought I would open issue here.

At https://github.com/krm1312/sb-1.2-websocket is a very small sample boot project that shows the issue.  You will need an external RabbitMQ (I think) to see it.

In the pom, it is currently set to run with 1.2.6.  It sends a 1000 messages in around 500ms.

```
...
2015-09-21 19:47:20.141  INFO 88077 --- [           main] o.s.c.support.DefaultLifecycleProcessor  : Starting beans in phase 2147483647
2015-09-21 19:47:20.142  INFO 88077 --- [           main] o.s.m.s.s.StompBrokerRelayMessageHandler : Starting...
2015-09-21 19:47:20.371  INFO 88077 --- [           main] o.s.m.s.s.StompBrokerRelayMessageHandler : Connecting "system" session to 127.0.0.1:61613
2015-09-21 19:47:20.446  INFO 88077 --- [           main] o.s.m.s.s.StompBrokerRelayMessageHandler : Started.
2015-09-21 19:47:20.463  INFO 88077 --- [eactor-tcp-io-1] reactor.net.netty.tcp.NettyTcpClient     : CONNECTED: [id: 0x17c08180, /127.0.0.1:53884 => /127.0.0.1:61613]
2015-09-21 19:47:20.496  INFO 88077 --- [eactor-tcp-io-1] o.s.m.s.s.StompBrokerRelayMessageHandler : "System" session connected.
2015-09-21 19:47:20.499  INFO 88077 --- [eactor-tcp-io-1] o.s.m.s.s.StompBrokerRelayMessageHandler : BrokerAvailabilityEvent[available=true, StompBrokerRelay[127.0.0.1:61613]]
2015-09-21 19:47:20.505  INFO 88077 --- [           main] s.b.c.e.t.TomcatEmbeddedServletContainer : Tomcat started on port(s): 8080 (http)
2015-09-21 19:47:20.508  INFO 88077 --- [           main] org.krm1312.test.App                     : Started App in 3.922 seconds (JVM running for 4.359)
Sending msg: 0
Sending msg: 1
Sending msg: 2
Sending msg: 3
Sending msg: 4
Sending msg: 5
...
Sent 1000 messages in 388.9 ms
```

If you edit the pom, and change parent to 1.3.0.M5 and flip the reactor dependencies at the bottom, it's lucky to send one message a second.

```
2015-09-21 20:01:41.684  INFO 93055 --- [           main] o.s.m.s.s.StompBrokerRelayMessageHandler : Starting...
2015-09-21 20:01:41.745  INFO 93055 --- [           main] o.s.m.s.s.StompBrokerRelayMessageHandler : Connecting "system" session to 127.0.0.1:61613
2015-09-21 20:01:41.875  INFO 93055 --- [           main] o.s.m.s.s.StompBrokerRelayMessageHandler : Started.
2015-09-21 20:01:41.889  INFO 93055 --- [eactor-tcp-io-1] r.io.net.impl.netty.tcp.NettyTcpClient   : CONNECTED: [id: 0x4a42a22b, /127.0.0.1:54139 => /127.0.0.1:61613]
2015-09-21 20:01:41.942  INFO 93055 --- [           main] s.b.c.e.t.TomcatEmbeddedServletContainer : Tomcat started on port(s): 8080 (http)
2015-09-21 20:01:41.945  INFO 93055 --- [eactor-tcp-io-1] o.s.m.s.s.StompBrokerRelayMessageHandler : "System" session connected.
2015-09-21 20:01:41.948  INFO 93055 --- [           main] org.krm1312.test.App                     : Started App in 4.343 seconds (JVM running for 4.767)
2015-09-21 20:01:41.951  INFO 93055 --- [eactor-tcp-io-1] o.s.m.s.s.StompBrokerRelayMessageHandler : BrokerAvailabilityEvent[available=true, StompBrokerRelay[127.0.0.1:61613]]
Sending msg: 0
Sending msg: 1
Sending msg: 2
Sending msg: 3
Sending msg: 4
Sending msg: 5
Sending msg: 6
Sending msg: 7
Sending msg: 8
Sending msg: 9
Sending msg: 10
Sending msg: 11
Sending msg: 12
Sending msg: 13
Sending msg: 14
Giving up after only sending 14 messages
Sent 14 messages in 11.93 s
```

My guess is it has something to do with the reactor upgrade.

Environment: OS X

**Comment by rstoyanchev** [comment]
I'll have a look.

**Comment by rstoyanchev** [comment]
It looks like a Reactor issue. I created a branch with a more focused example that uses only the Reactor TcpClient and it has the same behavior: https://github.com/rstoyanchev/sb-1.2-websocket/tree/client-app

This is the output from running ClientApp:
````
18:17:37.519 [main] DEBUG org.krm1312.test.ClientApp - Sending message: 0
18:17:38.521 [reactor-tcp-io-1] DEBUG r.i.n.impl.netty.tcp.NettyTcpClient - [id: 0x8b679ca7, /127.0.0.1:56796 => /127.0.0.1:61613] WRITE: 47B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 53 45 4e 44 0a 64 65 73 74 69 6e 61 74 69 6f 6e |SEND.destination|
|00000010| 3a 2f 74 6f 70 69 63 2f 66 6f 6f 0a 63 6f 6e 74 |:/topic/foo.cont|
|00000020| 65 6e 74 2d 6c 65 6e 67 74 68 3a 30 0a 0a 00    |ent-length:0... |
+--------+-------------------------------------------------+----------------+
18:17:38.522 [reactor-tcp-io-1] DEBUG r.i.n.impl.netty.tcp.NettyTcpClient - [id: 0x8b679ca7, /127.0.0.1:56796 => /127.0.0.1:61613] FLUSH
18:17:38.523 [main] DEBUG org.krm1312.test.ClientApp - Sending message: 1
18:17:39.528 [reactor-tcp-io-1] DEBUG r.i.n.impl.netty.tcp.NettyTcpClient - [id: 0x8b679ca7, /127.0.0.1:56796 => /127.0.0.1:61613] WRITE: 47B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 53 45 4e 44 0a 64 65 73 74 69 6e 61 74 69 6f 6e |SEND.destination|
|00000010| 3a 2f 74 6f 70 69 63 2f 66 6f 6f 0a 63 6f 6e 74 |:/topic/foo.cont|
|00000020| 65 6e 74 2d 6c 65 6e 67 74 68 3a 30 0a 0a 00    |ent-length:0... |
+--------+-------------------------------------------------+----------------+
18:17:39.528 [reactor-tcp-io-1] DEBUG r.i.n.impl.netty.tcp.NettyTcpClient - [id: 0x8b679ca7, /127.0.0.1:56796 => /127.0.0.1:61613] FLUSH
18:17:39.529 [main] DEBUG org.krm1312.test.ClientApp - Sending message: 2
18:17:40.529 [reactor-tcp-io-1] DEBUG r.i.n.impl.netty.tcp.NettyTcpClient - [id: 0x8b679ca7, /127.0.0.1:56796 => /127.0.0.1:61613] WRITE: 47B
````

There is a 1 second delay between the "main" thread that sends the message and the Netty thread "reactor-tcp-io-1" that actually writes.

**Comment by rstoyanchev** [comment]
https://github.com/reactor/reactor/issues/551.

**Comment by philwebb** [comment]
Thanks for the analysis @rstoyanchev!

**Comment by smaldini** [comment]
I'll need additional infos about JVM and JDK possibly but nevertheless I am looking at it right now (https://github.com/reactor/reactor/issues/551#issuecomment-143726233)

**Comment by rstoyanchev** [comment]
A fix is ready in Reactor 2.0.6 snapshots along with an additional change in Spring Framework 4.2.2 snapshots to adapt to a config method change in Reactor.
